File: 05_fp_challenges/09_merge_intervals/Solution09.java
// Merge Intervals  (LeetCode 56)
// ===============================
// Merge all overlapping intervals. Return them sorted by start.
//   merge([[1,3],[2,6],[8,10],[15,18]]) == [[1,6],[8,10],[15,18]]
//   merge([[1,4],[4,5]]) == [[1,5]]

import java.util.*;

public class Solution09 {

    static int[][] merge(int[][] intervals) {
        throw new UnsupportedOperationException("TODO");
    }

    // ---- tests ----
    public static void main(String[] args) {
        check(merge(new int[][]{{1,3},{2,6},{8,10},{15,18}}), new int[][]{{1,6},{8,10},{15,18}});
        check(merge(new int[][]{{1,4},{4,5}}),                new int[][]{{1,5}});
        check(merge(new int[][]{{1,4},{2,3}}),                new int[][]{{1,4}});
        check(merge(new int[][]{}),                           new int[][]{});
        System.out.println("all passed");
    }

    static void check(int[][] got, int[][] want) {
        if (!Arrays.deepEquals(got, want))
            throw new AssertionError("FAIL got " + Arrays.deepToString(got) + " want " + Arrays.deepToString(want));
        System.out.println("  PASS  " + Arrays.deepToString(got));
    }
}

File: 05_fp_challenges/09_merge_intervals/Solution09.hs
-- Merge Intervals  (LeetCode 56)
-- ===============================
-- Merge all overlapping intervals. Return them sorted by start.
--   merge [(1,3),(2,6),(8,10),(15,18)] == [(1,6),(8,10),(15,18)]
--   merge [(1,4),(4,5)] == [(1,5)]

module Main where

import Data.List (sortBy)
import Data.Ord (comparing)

merge :: [(Int, Int)] -> [(Int, Int)]
merge [] = []
merge intervals@(start : remaining) = reverse (last : merged)
  where
    (merged, last) = foldl (\(merged, prev) interval -> if fst interval <= snd prev then (merged, (fst prev, max (snd interval) (snd prev))) else (prev : merged, interval)) ([], start) (sortBy (comparing fst) intervals)

-- ---- tests ----
main :: IO ()
main = do
  check (merge [(1, 3), (2, 6), (8, 10), (15, 18)]) [(1, 6), (8, 10), (15, 18)]
  check (merge [(1, 4), (4, 5)]) [(1, 5)]
  check (merge [(1, 4), (2, 3)]) [(1, 4)]
  check (merge []) []
  putStrLn "all passed"
  where
    check got want
      | got == want = putStrLn ("  PASS  " ++ show got)
      | otherwise = error ("  FAIL  got " ++ show got ++ " want " ++ show want)
